cookbooks/anisotropic_viscosity/av_material.cc

```diff
@@ -0,0 +1,858 @@
+                }
+              n_dot[i] += Wn * n[j];
+            }
+          }
+        // make sure that n is a unit vector. for this to work,
+        // n needs to be a nonzero vector
+
+        Tensor<1,dim> n_new;
+        n_new = n + (n_dot * this->get_timestep());
+        Assert (n_new.norm() != 0, ExcInternalError());
+        n_new /= n_new.norm();
+        if (this->get_timestep()==0) //because at the start of the model timestep is 0
+        {
+
+        	n_dot=(n/n.norm())-n;
+        }
+        else {
+        	n_dot = (n_new-n)/ this->get_timestep();
+        	}
+          }
+        else
+        {
+        	n_dot=0;
+        }
+        //update  n[i] = in.composition[q][c_idx_n[i]] with adding to it n_dot*dt
+        for (unsigned int i=0; i<dim; ++i)
+            if (this->get_timestep()==0)
+            	out.reaction_terms[q][c_idx_n[i]] = n_dot[i];
+            else
+            	out.reaction_terms[q][c_idx_n[i]] = n_dot[i] * this->get_timestep();
+
+        if (n.norm() > 0.5)
+          {
+        	n/=n.norm();
+            SymmetricTensor<4,dim> Lambda;
+            for (unsigned int i=0; i<dim; ++i)
+              for (unsigned int j=0; j<dim; ++j)
+                for (unsigned int k=0; k<dim; ++k)
+                  for (unsigned int l=0; l<dim; ++l)
+                    Lambda[i][j][k][l] = 1./2. * (n[i]*n[k]*delta(l,j)
+                                                  + n[j]*n[k]*delta(i,l)
+                                                  + n[i]*n[l]*delta(k,j)
+                                                  + n[j]*n[l]*delta(i,k))
+                                         - 2*n[i]*n[j]*n[k]*n[l];
+
+              if (anisotropic_viscosity != nullptr)
+              {
+                  anisotropic_viscosity->stress_strain_directors[q] =dealii::identity_tensor<dim> ()
+                  - (1. - eta_0) * Lambda;
+                  SymmetricTensor<2,3> ViscoTensor;
+                  ViscoTensor[0][0]=anisotropic_viscosity->stress_strain_directors[q][0][0][0][0];
+                  ViscoTensor[0][1]=anisotropic_viscosity->stress_strain_directors[q][0][0][1][1];
+                  ViscoTensor[0][2]=anisotropic_viscosity->stress_strain_directors[q][0][0][0][1] * sqrt(2);
+                  ViscoTensor[1][0]=anisotropic_viscosity->stress_strain_directors[q][1][1][0][0];
+                  ViscoTensor[1][1]=anisotropic_viscosity->stress_strain_directors[q][1][1][1][1];
+                  ViscoTensor[1][2]=anisotropic_viscosity->stress_strain_directors[q][1][1][0][1] * sqrt(2);
+                  ViscoTensor[2][0]=anisotropic_viscosity->stress_strain_directors[q][0][1][0][0] * sqrt(2);
+                  ViscoTensor[2][1]=anisotropic_viscosity->stress_strain_directors[q][0][1][1][1] * sqrt(2);
+                  ViscoTensor[2][2]=anisotropic_viscosity->stress_strain_directors[q][0][1][0][1] * 2;
+
+                  Viscoeigenvalues[q] = eigenvalues(ViscoTensor);

```

Comment:
You don't actually need an array of these eigenvalues -- you really only need them one at a time. So declare
```
  const std::array<double,3> visco_eigenvalues = eigenvalues(ViscoTensor);
```
here. Then you don't need the array declaration above (as commented on).